x/dex/types/message_multi_hop_swap.go

```diff
@@ -70,17 +70,32 @@ func (msg *MsgMultiHopSwap) Validate() error {
 	if len(msg.Routes) == 0 {
 		return ErrMissingMultihopRoute
 	}
-
+	// set the expected tokenIn as the first token in the first route.
+	// if the route is empty, don't set tokenIn and let route length check error.
+	var tokenIn string
+	if len(msg.Routes[0].Hops) != 0 {
+		tokenIn = msg.Routes[0].Hops[0]
+	}
 	for _, route := range msg.Routes {
 		if len(route.Hops) < 2 {
 			return ErrRouteWithoutExitToken
 		}
 
+		if route.Hops[0] != tokenIn {
+			return sdkerrors.Wrapf(ErrInvalidDenom, "TokenIn mismatch in routes")
```

Comment:
Not really an `InvalidDenom` issue. I would either create a new error `ErrMultihopEntryTokensMismatch` or create a more general error that can be used for mismatch entry/exit tokens